File: program/marketing/MarketingImportData.py
import csv
import kvstore

class MarketingImportData:
    def __init__(self):
        self.store = kvstore.Client('http://localhost:5000')
        self.tabMarketing = "MARKETINGS"
        self.pathToCSVFile = "../../vagrant/projet/data/Marketing.csv"
        self.clientID = 1
        # récupérer la valeur de l'instance kvstore client, vérifier si elle est déjà initialisée
        try:
            self.store.get(self.tabMarketing)
        except Exception as e:
            print(f"Erreur lors de la récupération de l'instance kvstore client : {e}")
            self.store = kvstore.Client('http://localhost:5000')
            print("Nouvelle instance kvstore client initialisée")

    def main(self):
        try:
            self.init_marketing_tables_and_data()
        except Exception as e:
            print(e)

    def init_marketing_tables_and_data(self):
        self.drop_table_marketing()
        self.create_table_marketing()
        self.load_marketing_data_from_file()

    def display_result(self, result, statement):
        print("===========================")
        print("Résultat de l'opération :")
        print(result)

    def drop_table_marketing(self):
        statement = f"drop table IF EXISTS {self.tabMarketing}"
        self.execute_ddl(statement)

    def create_table_marketing(self):
        statement = (
            f"create table IF NOT EXISTS {self.tabMarketing} ("
            + "ID INTEGER,"
            + "AGE STRING,"
            + "SEXE STRING,"
            + "TAUX STRING,"
            + "SITUATIONFAMILIALE STRING,"
            + "NBENFANTSACHARGE STRING,"
            + "DEUXIEMEVOITURE STRING,"
            + "PRIMARY KEY (ID))"
        )
        self.execute_ddl(statement)

    def execute_ddl(self, statement):
        try:
            result = self.store.set(self.tabMarketing, (statement))
            self.display_result(result, statement)
        except Exception as e:
            print(f"Erreur lors de l'exécution de la commande DDL : {e}")

    def insert_a_marketing_row(self, age, sexe, taux, situationFamiliale, nbEnfantsAcharge, deuxiemeVoiture):
        try:
            marketingRow = {
                "ID": self.clientID,
                "AGE": age,
                "SEXE": sexe,
                "TAUX": taux,
                "SITUATIONFAMILIALE": situationFamiliale,
                "NBENFANTSACHARGE": nbEnfantsAcharge,
                "DEUXIEMEVOITURE": deuxiemeVoiture
            }
            result = self.store.set(self.tabMarketing, marketingRow)
            self.display_result(result, f"Insertion d'une ligne de marketing avec ID: {self.clientID}")
            self.clientID += 1
        except Exception as e:
            print(f"Erreur lors de l'insertion d'une ligne de marketing : {e}")

    def load_marketing_data_from_file(self):
        try:
            with open(self.pathToCSVFile, newline='', encoding='latin-1') as csvfile:
                reader = csv.reader(csvfile)
                for row in reader:
                    age, sexe, taux, situationFamiliale, nbEnfantsAcharge, deuxiemeVoiture = row
                    self.insert_a_marketing_row(age, sexe, taux, situationFamiliale, nbEnfantsAcharge, deuxiemeVoiture)
        except Exception as e:
            print(f"Erreur lors du chargement des données marketing depuis le fichier : {e}")

if __name__ == "__main__":
    marketing_import_data = MarketingImportData()
    marketing_import_data.main()

File: program/marketing/MarketingImportData.java
package marketing;

import oracle.kv.KVStore;
import java.util.*;
import oracle.kv.*;
import oracle.kv.StatementResult;
import oracle.kv.table.*;
import java.lang.Integer;
import java.io.*;

public class MarketingImportData {

    private final KVStore store;
    private final String TABLE_MARKETING = "MARKETING";
    private final String CSV_FILE_PATH = "../../vagrant/projet/data/Marketing.csv";
    private int clientId = 1;

    public static void main(String[] args) {
        try {
            MarketingImportData mark = new MarketingImportData(args);
            mark.initMarketingTablesAndData();
        } catch (RuntimeException e) {
            e.printStackTrace();
        }
    }

    MarketingImportData(String[] argv) {
        String storeName = "kvstore";
        String hostName = "localhost";
        String hostPort = "5000";
        store = KVStoreFactory.getStore(new KVStoreConfig(storeName, hostName + ":" + hostPort));
    }

    private void displayResult(StatementResult result, String statement) {
        System.out.println("===========================");
        if (result.isSuccessful()) {
            System.out.println("Statement was successful:\n\t" + statement);
            System.out.println("Results:\n\t" + result.getInfo());
        } else if (result.isCancelled()) {
            System.out.println("Statement was cancelled:\n\t" + statement);
        } else {
            if (result.isDone()) {
                System.out.println("Statement failed:\n\t" + statement);
                System.out.println("Problem:\n\t" + result.getErrorMessage());
            } else {
                System.out.println("Statement in progress:\n\t" + statement);
                System.out.println("Status:\n\t" + result.getInfo());
            }
        }
    }

    public void initMarketingTablesAndData() {
        dropTableMarketing();
        createTableMarketing();
        loadMarketingDataFromFile(CSV_FILE_PATH);
    }

    public void dropTableMarketing() {
        String statement = "DROP TABLE " + TABLE_MARKETING;
        executeDDL(statement);
    }

    public void createTableMarketing() {
        String statement = "CREATE TABLE " + TABLE_MARKETING + " (" +
                           "ID INTEGER," +
                           "AGE STRING," +
                           "SEXE STRING," +
                           "TAUX STRING," +
                           "SITUATIONFAMILIALE STRING," +
                           "NBENFANTSACHARGE STRING," +
                           "DEUXIEMEVOITURE STRING," +
                           "PRIMARY KEY (ID))";
        executeDDL(statement);
    }

    public void executeDDL(String statement) {
        TableAPI tableAPI = store.getTableAPI();
        System.out.println("****** Executing DDL: " + statement + " ********");
        try {
            StatementResult result = store.executeSync(statement);
            displayResult(result, statement);
        } catch (IllegalArgumentException | FaultException e) {
            System.out.println("Error executing statement:\n" + e.getMessage());
        }
    }

    private void insertMarketingRow(String age, String sexe, String taux, String situationFamiliale, String nbEnfantsAcharge, String deuxiemeVoiture) {
        System.out.println("***** Inserting Row in Marketing Table *****");

        try {
            TableAPI tableAPI = store.getTableAPI();
            Table marketingTable = tableAPI.getTable(TABLE_MARKETING);
            Row marketingRow = marketingTable.createRow();

            marketingRow.put("ID", clientId++);
            marketingRow.put("age", age);
            marketingRow.put("sexe", sexe);
            marketingRow.put("taux", taux);
            marketingRow.put("situationFamiliale", situationFamiliale);
            marketingRow.put("nbEnfantsAcharge", nbEnfantsAcharge);
            marketingRow.put("deuxiemeVoiture", deuxiemeVoiture);

            tableAPI.put(marketingRow, null, null);
        } catch (IllegalArgumentException | FaultException e) {
            System.out.println("Error inserting row:\n" + e.getMessage());
        }
    }

    void loadMarketingDataFromFile(String marketingDataFileName) {
        System.out.println("***** Loading Data From File: " + marketingDataFileName + " *****");

        try (BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(marketingDataFileName)))) {

            br.readLine();

            String line;
            while ((line = br.readLine()) != null) {
                StringTokenizer tokenizer = new StringTokenizer(line, ",");
                List<String> marketingRecord = new ArrayList<>();
                while (tokenizer.hasMoreTokens()) {
                    marketingRecord.add(tokenizer.nextToken());
                }

                insertMarketingRow(marketingRecord.get(0), marketingRecord.get(1), marketingRecord.get(2),
                                   marketingRecord.get(3), marketingRecord.get(4), marketingRecord.get(5));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}